Correctly drop router receiver after shutdown

spacekookie opened the pull request:
This commit (potentially) fixes #3105.  Because the router run-loop does not own the Router, it is not dropped after the loop terminates.

Note: the commit isn't signed because I forgot my Yubikey at home.  I will add a signature later.

twittner commented on implementations/rust/ockam/ockam_node/src/router/mod.rs:
```diff
@@ -269,11 +276,13 @@ impl Router {
                 Ok(should_break) => {
                     if should_break {
+                        // We drop the receiver end here
+                        self.receiver.take();
```
Rather than dropping the receiver I recommend calling `Receiver::close` instead. This way there is no need to have it as an `Option`al field in `Router`.

spacekookie commented:
Oh @twittner `Receiver::close` isn't implemented in `no_std`?? @antoinevg 

I can just go back to wrapping it in an Option but this is a bit strange

spacekookie committed:
fix(rust): drop router receiver after shutdown

This commit (potentially) fixes #3105.  Because the router run-loop
does not own the Router, it is not dropped after the loop terminates

antoinevg commented:
> Oh @twittner `Receiver::close` isn't implemented in `no_std`?? @antoinevg
> 
> I can just go back to wrapping it in an Option but this is a bit strange

We can probably switch to `futures::channel` on `no_std` rather than using our own impl. I'll take a look at this and open a PR if it's do-able.

spacekookie commented:
I just changed the code to be as it was before, let's just merge this option first then so we can see if it makes a difference in how often we have CI failures

twittner commented:
> [...] `Receiver::close` isn't implemented in `no_std`?? [...]

For the time being it is sufficient to add a stub to https://github.com/build-trust/ockam/blob/2ea26c64036e61669181f17c92b0906a6a434a79/implementations/rust/ockam/ockam_executor/src/channel.rs#L139 as I did here: https://github.com/build-trust/ockam/commit/d6b68e48d25167b9d68f6a81bd9982770536115a

antoinevg commented:
> We can probably switch to `futures::channel` on `no_std` rather than using our own impl. I'll take a look at this and open a PR if it's do-able.

Ah never mind, `futures::channel::mpsc` is only available on `std`.

Just adding a stub for `close()` is fine as the executor is single-threaded so there's no need to notify anything.

spacekookie commented:
@twittner either merge this PR or close it and open a PR with your solution. I'm not rebasing it again
